### PR TITLE
[None][feat] Enable EPLB for trtllm-gen and cutlass backend

### PR DESCRIPTION
**Pull request by dongxuy04**
## Summary by CodeRabbit

* **New Features**
  * Added load balancer support for multiple Mixture of Experts backends, enabling improved token routing and distribution.
  * Introduced auxiliary CUDA stream support for optimized load balancer operations.
  * Enhanced memory management with CPU tensor pageout hints for efficient memory handling.
  * Extended model architecture support for GptOssForCausalLM.

* **Tests**
  * Expanded test coverage with parameterized MoE backend configurations and online load balancer scenarios.

## Description

Enable EPLB for trtllm-gen and cutlass backend:
- Refactor EPLB logic to base MoE class
- Add EPLB weight loading for more quant methods of trtllm-gen backend like W4A8MXFP4MXFP8TRTLLMGenFusedMoEMethod.
- Enable online EPLB for trtllm-gen and cutlass backend
- Optimize memory usage by unmap some weights that might not used again
- Explicitly mark EPLB unsupport and unverified quant methods
- Add and update some tests.

Known issues:

For now, there might be some issues when using EPLB backend with TRTLLM-Gen backend and number of expert per group is larger than 32. The reason is that our current TRTLLM-Gen fused routing kernel doesn't support that as no models using that. But when using EPLB with redundant experts like EPLB288 for DeepSeek-R1, it seems that we have more than that. In thant case, ideally, we should not run the router inside TRTLLM-Gen, but since there are some issues with AutoTuner when using topk input in that case, it is not enabled now.

Test with GPT-OSS 120B on 4 GB200 nodes (16 GPU) with EP=16, BS=512, trtllm-bench IFB mode.

Without EPLB:
 0: ===========================================================
 0: = PERFORMANCE OVERVIEW 
 0: ===========================================================
 0: Request Throughput (req/sec):                     214.6129
 0: Total Output Throughput (tokens/sec):             211175.4890
 0: Total Token Throughput (tokens/sec):              441672.5608
 0: Total Latency (ms):                               38171.0446
 0: Average request latency (ms):                     35473.2092
 0: Per User Output Throughput [w/ ctx] (tps/user):   27.7429
 0: Per GPU Output Throughput (tps/gpu):              13198.4681

With 144 slot EPLB (16 redundant experts):
 0: ===========================================================
 0: = PERFORMANCE OVERVIEW 
 0: ===========================================================
 0: Request Throughput (req/sec):                     224.4187
 0: Total Output Throughput (tokens/sec):             220824.1976
 0: Total Token Throughput (tokens/sec):              461852.7904
 0: Total Latency (ms):                               36503.1961
 0: Average request latency (ms):                     33789.7656
 0: Per User Output Throughput [w/ ctx] (tps/user):   29.1216
 0: Per GPU Output Throughput (tps/gpu):              13801.5124

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: cpp/tensorrt_llm/thop/mxFp4BlockScaleMoe.cpp [hunk 1]
```diff
@@ -200,8 +200,8 @@ torch::Tensor dtype_mxe2m1_block_scale_moe_runner(torch::optional<torch::Tensor>
             max_num_padded_tokens, args.hidden_size, btg::dtypeGetNumBits(args.mDtypeOut));
     at::Tensor total_num_padded_tokens
         = at::empty({}, at::TensorOptions().device(routing_device).dtype(at::ScalarType::Int));
-    at::Tensor expanded_idx_to_permuted_idx
-        = at::detail::empty_cuda({args.num_tokens * args.top_k}, at::ScalarType::Int, routing_device, std::nullopt);
+    at::Tensor expanded_idx_to_permuted_idx = at::full(
```
Comment (nekorobov) [inline cpp/tensorrt_llm/thop/mxFp4BlockScaleMoe.cpp]: Have you measured the perf impact of this? Also, do you need to do the same for other accuracies. I.e. DS FP8, FP8, NVFP4?

File: tensorrt_llm/_torch/modules/fused_moe/fused_moe_trtllm_gen.py [hunk 2]
```diff
@@ -326,6 +327,34 @@ def forward_impl(
...
+            if self.layer_load_balancer:
+                # Determine if this is first/last call (TRTLLMGenFusedMoE doesn't use chunking)
+                is_first_call = self.repeat_idx == 0
+                is_last_call = self.repeat_idx == self.repeat_count - 1
+
+                # Start GPU stage for first call
+                self._load_balancer_start_wait_gpu_stage(is_first_call)
+                self._load_balancer_done_wait_gpu_stage(is_first_call)
+
+                # Update load balancer statistics (TRTLLMGenFusedMoE doesn't use MNNVL)
```
Comment (nekorobov) [inline tensorrt_llm/_torch/modules/fused_moe/fused_moe_trtllm_gen.py]: Support for MNNVL for TRTLLM Gen was merged, iirc

Comment (dongxuy04) [reply tensorrt_llm/_torch/modules/fused_moe/fused_moe_trtllm_gen.py]: Thanks for helping point this out. Added MNNVL support of ELPB for TRTLLM-Gen backend.

File: tensorrt_llm/_torch/modules/fused_moe/fused_moe_cutlass.py [hunk 3]
```diff
@@ -665,37 +702,42 @@ def _reducescatter_or_allreduce(x_, idx):
             # Postpone reduce-scatter/all-reduce to the next iteration to achieve better overlap
             for idx_chunk, (x, router_logits) in enumerate(
                     zip(x_list, router_logits_list)):
-
-                if idx_chunk % 2 == 0:
-                    with torch.cuda.stream(self.aux_stream):
+                if not (self.moe_alltoall_backend == "mnnvllatency"):
```
Comment (xxi-nv) [inline tensorrt_llm/_torch/modules/fused_moe/fused_moe_cutlass.py]: Are you intend to avoid the multi-chunk when using ALLTOALL?
If yes, maybe you need to extend the condition.

File: tensorrt_llm/_torch/modules/fused_moe/fused_moe_trtllm_gen.py [hunk 4]
```diff
@@ -348,6 +351,33 @@ def forward_impl(
             if token_final_scales is not None:
                 token_final_scales = token_final_scales.to(torch.bfloat16)
 
+            # Apply load balancer routing if available
+            if self.layer_load_balancer:
+                # Determine if this is first/last call (TRTLLMGenFusedMoE doesn't use chunking)
+
+                # Start GPU stage for first call
+                self._load_balancer_start_wait_gpu_stage(is_first_call)
```
Comment (xxi-nv) [inline tensorrt_llm/_torch/modules/fused_moe/fused_moe_trtllm_gen.py]: self._load_balancer_start_wait_gpu_stage(is_first_call) could be called before the routing_method.apply().

File: tensorrt_llm/_torch/modules/fused_moe/quantization.py [hunk 5]
```diff
@@ -156,6 +156,23 @@ class FusedMoEMethodBase(ABC):
...
+        ):
+            raise NotImplementedError(
+                f'{cls.__name__} doesn\'t support online EPLB now')
+
+    @classmethod
+    def _online_eplb_not_verified(cls, module):
+        if hasattr(
+                module, "layer_load_balancer"
+        ) and module.layer_load_balancer and module.layer_load_balancer.need_load_shared_weights(
+        ):
```
Comment (xxi-nv) [inline tensorrt_llm/_torch/modules/fused_moe/quantization.py]: How about create a unify helper to judge the following condition which is called multi-times.
hasattr(
                module, "layer_load_balancer"
        ) and module.layer_load_balancer and module.layer_load_balancer.need_load_shared_weights(
        )

Comment (dongxuy04) [reply tensorrt_llm/_torch/modules/fused_moe/quantization.py]: Refactored `need_load_shared_weights` as class method to reuse that.

**Review by xxi-nv — CHANGES_REQUESTED** [review]

File: tensorrt_llm/_torch/modules/fused_moe/quantization.py [hunk 6]
```diff
@@ -1924,19 +1979,23 @@ def load_expert_w2_weight(self, module: torch.nn.Module,
...
-                            non_blocking=True)
+        dst_w2_weight_gpu.copy_(processed_w2_weight.view(
+            dst_w2_weight_gpu.dtype),
+                                non_blocking=dst_on_gpu)
+
+        if not dst_on_gpu:
+            dst_w2_weight.copy_(dst_w2_weight_gpu)
 
     def load_expert_w3_w1_weight_scale_nvfp4(
             self,
```
[the remark this answers is not in the document]
Comment (dongxuy04) [reply tensorrt_llm/_torch/modules/fused_moe/quantization.py]: For some quantization methods that don't need GPU transformation or don't use partial or full `dst_xxx_weight` as GPU buffer during transformation, we don't need separate version of that.

**Review by xxi-nv — APPROVED** [review]

**Review by bobboli — APPROVED** [review]

**Review by syuoni — APPROVED** [review]

File: tests/integration/test_lists/qa/llm_function_core.txt
Comment (yuxianq) [inline tests/integration/test_lists/qa/llm_function_core.txt]: Why do we remove `accuracy/test_disaggregated_serving.py::TestGPTOSS::test_auto_dtype[True]` and `accuracy/test_disaggregated_serving.py::TestGPTOSS::test_auto_dtype[False]`?

File: tensorrt_llm/_torch/modules/fused_moe/moe_load_balancer.py [hunk 7]
```diff
@@ -160,6 +240,16 @@ def share_host_tensor_with_shape(self, expert_id: int, name: str,
     def align_size(size: int):
         return (size + 256 - 1) // 256 * 256
 
+    def add_raw_host_weight_for_unmap(self,
+                                      raw_weight_tensors: List[torch.Tensor]):
+        """
+        Add a raw weight (mmapped Tensor) to HostMoeTensorSharer for later madvise to save host memory.
```
Comment (yuxianq) [inline tensorrt_llm/_torch/modules/fused_moe/moe_load_balancer.py]: "madvise" -> "advise"

Comment (dongxuy04) [reply tensorrt_llm/_torch/modules/fused_moe/moe_load_balancer.py]: Updated, thanks.

**Review by yuxianq — APPROVED** [review]

File: tensorrt_llm/_torch/modules/fused_moe/quantization.py [hunk 8]
```diff
@@ -1924,19 +1979,23 @@ def load_expert_w2_weight(self, module: torch.nn.Module,
         epilogue_tile_m = 128
 
         # Keep weights in device buffer
-        dst_w2_weight.copy_(w2_weight_shard.view(dst_w2_weight.dtype),
-                            non_blocking=True)
+        dst_w2_weight_gpu.copy_(w2_weight_shard.view(dst_w2_weight_gpu.dtype),
+                                non_blocking=dst_on_gpu)
```
[the remark this answers is not in the document]
Comment (dongxuy04) [reply tensorrt_llm/_torch/modules/fused_moe/quantization.py]: Same as above.
